File: 二叉树/完全二叉树插入器.go
package 二叉树

type CBTInserter struct {
	queue []*TreeNode //采用数组储存二叉树，方便检索一个节点的父节点
}

func Constructor0(root *TreeNode) CBTInserter {
	//层序遍历root二叉树，构建一个root二叉树的数组储存形式
	tmp := []*TreeNode{root}
	var queue []*TreeNode
	for len(tmp) != 0 {
		cur := tmp[0]
		tmp = tmp[1:]
		queue = append(queue, cur)
		if cur.Left != nil {
			tmp = append(tmp, cur.Left)
		}
		if cur.Right != nil {
			tmp = append(tmp, cur.Right)
		}
	}
	return CBTInserter{queue}
}

func (this *CBTInserter) Insert(val int) int {
	n := len(this.queue) //n就是新插入的节点在数组中的下标
	tmp := &TreeNode{val, nil, nil}
	this.queue = append(this.queue, tmp)
	father := this.queue[(n-1)/2]
	if father.Left == nil {
		father.Left = tmp
	} else {
		father.Right = tmp
	}
	return father.Val
}

func (this *CBTInserter) Get_root() *TreeNode {
	return this.queue[0]
}

File: 二叉树/完全二叉树插入器.java
package 二叉树;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;

//剑指offer
//利用数组储存完全二叉树，可以快速从子节点追溯到父节点（前提是一定要是完全二叉树哦）
public class 完全二叉树插入器 {

}

class CBTInserter {
    private List<TreeNode> tree;

    public CBTInserter(TreeNode root) {
        tree = new ArrayList<TreeNode>();
        if (root == null) return;
        var queue = new ArrayDeque<TreeNode>();
        queue.add(root);
        while (!queue.isEmpty()) {
            var cur = queue.poll();
            tree.add(cur);
            if (cur.left != null) queue.add(cur.left);
            if (cur.right != null) queue.add(cur.right);
        }
    }

    public int insert(int v) {
        var newt = new TreeNode(v);
        tree.add(newt);
        int fatherindex = (tree.size() - 1 - 1) / 2;
        if (tree.get(fatherindex).left == null) {
            tree.get(fatherindex).left = newt;
        } else {
            tree.get(fatherindex).right = newt;
        }
        return tree.get(fatherindex).val;
    }

    public TreeNode get_root() {
        return tree.get(0);
    }
}

